### ISSUE TITLE
Discussion: Improving Google Lighthouse scores

### ISSUE THREAD

**Issue by jamesericdavidson** [526avijitgupta/gokarna#250]
## Problem Statement

The README mentions the [Google Lighthouse score for exampleSite](https://github.com/526avijitgupta/gokarna?tab=readme-ov-file#performance). Without context, the user could believe that Gokarna's performance is "baked in", and requires no action on their part.

Moreover, Lighthouse continues to identify minor issues with Gokarna, which can be either avoided in source code, or by virtue of how the site is deployed.

Please read and discuss below if this is of interest. :point_down: 

## User sites

Unless we direct users to evaluate their site with Lighthouse (and highlight the benefits of doing so), some site configurations will be less than ideal.

e.g. Users who do not resize images, or use "modern" formats (such as WebP), are penalised.

Suggestion: Some site-level changes are easy to make statically (i.e. without the use of CDNs, load balancers, and other at-scale deployment mechanisms), and should be recommended in the Gokarna docs. 

_:wave: I can pick this up after #243._

## exampleSite

When reviewing the [Lighthouse score for exampleSite](https://pagespeed.web.dev/analysis/https-gokarna-hugo-netlify-app/w90db0qrkg?form_factor=desktop), the following improvements are highlighted:

![image](https://github.com/user-attachments/assets/45dbbbfa-9b58-4b81-8272-a01c5ee77e18)

![image](https://github.com/user-attachments/assets/ad1d50d1-e0d8-495b-87b8-c6f6fe87c030)

![image](https://github.com/user-attachments/assets/01d07a79-2b79-4e8b-a2ec-b516bb6f7e83)

![image](https://github.com/user-attachments/assets/a7b86855-994d-4cc5-b0a8-2fd6d28a32fa)

Can we identify which of these are addressable:

- In source code?
- Via site deployment?

**Comment by jamesericdavidson** [comment]
We could also add:

- [Minification](https://gohugo.io/hugo-pipes/minification/)

**Comment by 526avijitgupta** [comment]
Hi @jamesericdavidson , going through the lighthouse score page I see we do still have a 100% score in performance.  Although we have 92 as our lowest score in the SEO category, which arguably is the more important one: 
<img width="505" alt="image" src="https://github.com/user-attachments/assets/17aabe9a-41b9-4d8b-9366-6a9bd4ce4a5f">

I feel like we should separate out the SEO part (and perhaps work on it first?). Therefore, I've created [a new issue](https://github.com/gokarna-theme/gokarna-hugo/issues/259) specifically for that topic. Would make it easier to track

**Comment by jamesericdavidson** [comment]
@526avijitgupta I think the word 'performance' was a poor choice on my part - I wanted this issue to begin addressing the red and yellow checks in all four sections, essentially making this an overarching discussion or Epic. Breaking this out is a good idea

**Comment by 526avijitgupta** [comment]
> Can we identify which of these are addressable:
> 
> In source code?
> Via site deployment?

Regarding this. I'll go through the open points one by one and categorize them. Hopefully this weekend
But if you get the time and want to work on it then feel free to do so :)

**Comment by jamesericdavidson** [comment]
IMO we should do as much as possible in Hugo (e.g. minification, even when that option is provided by web hosts/CDNs) before covering external options (such as compression).

e.g. The docs should prioritise Hugo and theme-specific functionality

**Comment by jamesericdavidson** [comment]
> I'll go through the open points one by one and categorize them.

@526avijitgupta Accessibility and SEO shouldn't need to be enumerated once merged: #260 #261 

Performance and Best Practices need looking at. Hopefully Best Practices is an easy fix

![image](https://github.com/user-attachments/assets/50d4f8ca-c4f3-42cf-b493-c345e8e0f36e)

**Comment by 526avijitgupta** [comment]
@jamesericdavidson I just re-ran the lighthouse audit and SEO is now at 100! 🎉 

<img width="997" alt="image" src="https://github.com/user-attachments/assets/2829e6a1-07a6-49bf-bed8-dab6dc2fa203">


But Accessibility is at 90 :/

<img width="1017" alt="image" src="https://github.com/user-attachments/assets/a3017ac7-c0ea-4a51-8b85-7258546346d5">

**Comment by jamesericdavidson** [comment]
@526avijitgupta interesting that these can't be worked on in isolation, I'll have another look

**Comment by jamesericdavidson** [comment]
This is caused by the element's `role`, which is needed to pass the SEO audits. Investigating now.

Good ARIA documentation:

https://accessibilityinsights.io/info-examples/web/aria-toggle-field-name/

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/switch_role

**Comment by jamesericdavidson** [comment]
Naively fixed: #265

**Comment by jamesericdavidson** [comment]
[New analysis](https://pagespeed.web.dev/analysis/https-gokarna-hugo-netlify-app/k71jt74rpe?form_factor=desktop) looks good for desktop:

![image](https://github.com/user-attachments/assets/a03aaf3e-4572-483c-94f5-5279a246492f)

![image](https://github.com/user-attachments/assets/98a5d5f6-ecb6-4092-90ec-e743997e4c51)

![image](https://github.com/user-attachments/assets/5394a52f-89be-4929-bc82-28cd49f5ed8f)

Need to expand mobile accessibility to hamburger menu:

![image](https://github.com/user-attachments/assets/ddd280a9-8a84-49fc-91a2-0beef1701edb)

**Comment by jamesericdavidson** [comment]
Syntax highlighting (code blocks) can trigger a Lighthouse audit:

![image](https://github.com/user-attachments/assets/180f2f59-89be-4027-851c-e982e56f5cd8)

I naively tried [`hr_high_contrast`](https://xyproto.github.io/splash/docs/hr_high_contrast.html), but no dice.

**Comment by jamesericdavidson** [comment]
@526avijitgupta I hope we can claw back some nebulous mobile performance by addressing audits in code.

Other than that, the strict px CSS rules for the avatar are outstanding, and perhaps others like the syntax highlighting (which are user-configurable, so the best we could do is identify/create themes which pass the audit).

**Comment by jamesericdavidson** [comment]
As above, I believe triaging and highlighting pitfalls in user sites should come *after* baking in everything we can (syntax highlighting being an odd mix of both)

**Comment by jamesericdavidson** [comment]
@526avijitgupta I've condensed the tasks discussed into the OP - please edit more in as appropriate, or leave a comment telling me about them.

Creating a PR now to pass more audits on more exampleSite pages

**Comment by 526avijitgupta** [comment]
Thanks @jamesericdavidson Mostly looks good to me. A few comments:

1. I would personally say we can skip addressing the color ratio for syntax highlighting. @yashmehrotra what do you think?
> Background:foreground colour ratio is insufficient:
> Syntax highlighting
> .post-date (dark mode)

2. Let's add a bullet point to the To-Do to update the screenshot of Desktop Lighthouse score in the README. It has become outdated with the new improvements :)

**Comment by JimOswick** [comment]
Just re-ran the Lighthouse test - we're still producing the same results on Desktop:

<img width="1023" height="1075" alt="Image" src="https://github.com/user-attachments/assets/d7e759c5-bb96-429d-9587-9e41d3d869ac" />

Closing this as it's rather stale, and there are no real gains to be made (we previously chose to ignore the drop in accessibility score due to code snippets lacking contrast - no sure which issue/PR this was discussed in)
